### ISSUE TITLE
Acquisition function suggests good points but never converges to the optimum

### ISSUE THREAD

**Issue by A-J-F-Mackintosh** [bayesian-optimization/BayesianOptimization#533]
Hi,

I am interested in apply Bayesian Optimization on a likelihood function with 10 parameters where other algorithms fail to converge to the optimum in a reasonable number of iterations. For example, Controlled Random Search cannot find the optimum even with 10,000 function evaluations.

I have a simpler version of the problem with only four parameters. This simpler version can be optimised with CRS + Nelder-Mead in 1500 + 500 function evaluations. I hoped to first apply Bayesian Optimization to this simpler problem before moving on to the higher dimension one.

My optimisation code looks like this (note that I have a method for providing good initial guesses and I use this to create some initial points before applying the GP):

```
acq = acquisition.UpperConfidenceBound(kappa=0.1, exploration_decay=0.95)

optimizer = BayesianOptimization(
		f=None, 
		acquisition_function=acq, 
		pbounds=pbounds, 
		verbose=2
		)

optimizer.set_gp_params(normalize_y=True, 
		kernel=Matern(length_scale=np.ones(dims)), n_restarts_optimizer=3, alpha=1e-4)

# generate some initial guesses rather than random points
for i in range(initial_guesses):
		next_point = func_for_initial_guesses()
		params = [next_point[key] for key in next_point]
		target = func(params)
		optimizer.register(params=next_point, target=target)

# let the GP model suggest new points
for i in range(max_iter - initial_guesses):
		next_point = optimizer.suggest()
		params = [next_point[key] for key in next_point]
		target = func(params)
		optimizer.register(params=next_point, target=target)
```

Below are the iterations where a new best point is found, and the target value of that point:

```
iter | target
-------------
0      -9227908.792
1      -7201969.103
3      -7197259.683
5      -7192594.252
6      -7190634.402
13     -7188447.419

##### from now on use GP to suggest points ####

31     -7188382.689
59     -7188045.823
171    -7187904.424
198    -7187828.195
```

As you can see, the GP model does not lead to much progress. I had the stop running the model at 575 evaluations, as beyond this point generating new suggestions becomes very slow. By observing the parameter suggestions I could see that they are often 'good' but they do not converge towards the global optimum (which has a target value of `-7187778.647`).

It's have tried changing the acquisition function (as well as kappa / xi) and it does not seem to improve the performance. My guess is that the predictions of the GP are accurate enough to generate 'good' suggestions, but not accurate enough to find the global optimum.

I am new to Bayesian optimisation and would be interested in any suggestions for improving performance. I have some a priori knowledge about the likelihood function (for example which parameters have the greatest effect on the target value and which have correlated errors when estimated) but I do not know how to include these information in the GP model. My understanding is that the hyper-parameters of the GP become tuned throughout the optimisation, but perhaps this is not working well in my case.

Any suggestions would be really appreciated.

Best,

Alex

**Comment by till-m** [comment]
Hey Alex,

it's a bit hard to diagnose what exactly is happening here without having access to any code or a better idea of what kind of problem you're trying to optimize.

A few things to consider:
- I would suggest you monitor the length-scale of the GP kernel during to optimization (it should be accessible via `optimizer._gp.kernel.length_scale` or a variant thereof). I suspect it collapses during the fitting.
- You have very few restarts of the optimizer, which is probably the cause of the collapse.
- Are your input variables of roughly the same magnitude? This can also cause problems with the fitting process.
- Are your targets approximately normally distributed? (`normalize_y=True` only shifts/scales the data to zero-mean and unit-variance)

NB:
- Your `kappa` value is quite small, which leads to a very greedy acquisition movement.
- When you say "can be optimized", what exactly do you mean? I assume that this is a continuous problem, so expecting the exact solution might be a bit much, if that's what you want.
- 2000 function evaluations also seems a lot for a four-dimensional problem though this of course depends on the smoothness of the function.

I hope this helps to some degree.

**Comment by A-J-F-Mackintosh** [comment]
Hi @till-m,

Thanks for the quick response.

> it's a bit hard to diagnose what exactly is happening here without having access to any code or a better idea of what kind of problem you're trying to optimize.

Apologies for the lack of detail. I am doing a maximum likelihood search on a population genetics model. The parameters have very different orders of magnitude (from 1e5 down to 1e-9). It is a continuous likelihood surface, although due to some rounding in the likelihood function it is possible that extremely similar points would give identical target values.

> I would suggest you monitor the length-scale of the GP kernel during to optimization (it should be accessible via optimizer._gp.kernel.length_scale or a variant thereof). I suspect it collapses during the fitting.

I tried printing `optimizer._gp.kernel.length_scale` after each function evaluation in the suggest, register loop
```
next_point = optimizer.suggest()
params = [next_point[key] for key in next_point]
target = func(params)
print(optimizer._gp.kernel.length_scale)
optimizer.register(params=next_point, target=target)
```
but this gives the same output each time `[1. 1. 1. 1.]`, which is the value I initiated with. Does this suggest that the length scale is not being tuned?

> You have very few restarts of the optimizer, which is probably the cause of the collapse.

I have increased this to `n_restarts_optimizer=30` and I think it improved performance.

> Are your input variables of roughly the same magnitude? This can also cause problems with the fitting process.

They are on very different magnitudes (see above). I can trying rescaling these to between 0 and 1, where 0 is the global lower bound and 1 the upper. I also expected some of these variables to have a marginal effect on the target that is on a log scale, rather than linear. So I can log transform these.

> Are your targets approximately normally distributed? (normalize_y=True only shifts/scales the data to zero-mean and unit-variance)

For the initial points? I think they likely have a very long tail, as some points have terrible target values. 

> Your kappa value is quite small, which leads to a very greedy acquisition movement.

Yes, I have now found that `acquisition.UpperConfidenceBound(kappa=1.96, exploration_decay=0.999)` gives slightly better behaviour.

> When you say "can be optimized", what exactly do you mean? I assume that this is a continuous problem, so expecting the exact solution might be a bit much, if that's what you want.

My goal is indeed to find the exact solution, or at least values very close to it. Perhaps my expectations are a bit too high... 

> 2000 function evaluations also seems a lot for a four-dimensional problem though this of course depends on the smoothness of the function.

I think the likelihood surface has quite many local minima (or at least ridges where the likelihood is high and flat, but not maximal), and the effect of the parameters on the target value are highly dependent on one another. It is easy to obtain a 'good' local solution in a few dozen iterations, but finding the global maximum consistently with CRS + Nelder-Mead does seem to take a couple of thousand evaluations.

Thanks again for the feedback. I am particularly interested in knowing whether the length scales are being tuned at all, as perhaps that is the issue. Is my print statement above correct, or am I just retrieving the parameter setting rather than the final value within the model?

Best,

Alex

**Comment by till-m** [comment]
Hi Alex,

first of all, what version of this library are you using?

> have very different orders of magnitude (from 1e5 down to 1e-9)

\+

> They are on very different magnitudes (see above). I can trying rescaling these to between 0 and 1, where 0 is the global lower bound and 1 the upper. I also expected some of these variables to have a marginal effect on the target that is on a log scale, rather than linear. So I can log transform these.

These are such extreme magnitudes that I think you would run into problems generally as they are outside the [bounds on the length scale](https://scikit-learn.org/dev/modules/generated/sklearn.gaussian_process.kernels.Matern.html), which are by default given as `(1e-5, 1e5)`. I would recommend rescaling to have features in `[0, 1]` as you mentioned. The log transform might also be a good idea -- the Matern kernel you're using is stationary, so only the difference between points is relevant.

> this gives the same output each time [1. 1. 1. 1.], which is the value I initiated with. Does this suggest that the length scale is not being tuned?

This is indeed concerning and it's unclear to me why this would happen. I currently don't have access to a working python environment, so I cannot have a closer look. It could be related to the bounds problem above^.

> I have increased this to n_restarts_optimizer=30 and I think it improved performance.

This is odd, as the only parameter being fit is the lengthscale, if I'm not mistaken...

> Is my print statement above correct, or am I just retrieving the parameter setting rather than the final value within the model?

I remember that it is possible to read out the actual length scales in some way similar to the command I stated, but I don't remember if it is exactly this command (maybe you could have a look at what properties the gp & the gp kernel have). I unfortunately currently don't have access to a working python environment, so I can't really help too much.

At present I'm thinking that it could either be that a) this is not the right way to read out the length-scale, b) this is the right way, but somehow the fitting does not happen (this is unlikely, as the library does handle this case), c) the fitting happens, but due to some problem (e.g. the extreme magnitudes of the features), [1 1 1 1] just ends up being the maximizer of the likelihood function.

I would suggest you start by re-scaling the features and then check again if the length-scale changes.

Apologies for not being able to help more, but please let me know what you have tried and what the outcome was.

**Comment by till-m** [comment]
Hey Alex,

I checked -- it's actually `optimizer._gp.kernel_.length_scale` and not `optimizer._gp.kernel.length_scale`. Let me know if this changes things ;)

Apologies!

**Comment by till-m** [comment]
Hey Alex, I'm assuming the problem has been solved. Feel free to re-open or make a new issue if you encounter more trouble.
